### PR TITLE
[TCling] Add method to get using declarations of namespaces

### PR DESCRIPTION
**Pull request by etejedor**
When updating `cppyy-backend` to v1.7.3 in PyROOT experimental, a compilation error appears because of a missing functionality in `TCling`, in particular a method to get using declarations of namespaces.

This functionality was added to Cppyy's ROOT recently, and a patch was created for it:
https://bitbucket.org/wlav/cppyy-backend/src/59a4a747d03c6e94eaaa28062daf83104334f27d/cling/patches/using_directives.diff

This PR proposes to add such functionality to mainstream ROOT. 

Inviting @Axel-Naumann and @wlav to the party.

### CODE REVIEW

**Comment by wlav** [comment]
Background is here:

https://bitbucket.org/wlav/cppyy/issues/61/using-statement-within-a-namespace-does

**Comment by wlav** [comment]
Note that in the same general area, there are several fixes to Cling, incl. using of base class methods and template aliases.

File: core/metacling/src/TClingClassInfo.cxx [hunk 1]
```diff
@@ -634,6 +634,33 @@ ptrdiff_t TClingClassInfo::GetBaseOffset(TClingClassInfo* base, void* address, b
    return binfo.Offset(address, isDerivedObject);
 }
 
+std::vector<std::string> TClingClassInfo::GetUsingNamespaces()
+{
+   // Find and return all 'using' declarations of namespaces.
+   std::vector<std::string> res;
+
+   R__LOCKGUARD(gInterpreterMutex);
+
```
Comment (vgvassilev) [inline core/metacling/src/TClingClassInfo.cxx]:
```suggestion
if (!isa<DeclContext>(DC))
  return {};
```

File: core/metacling/src/TClingClassInfo.cxx [hunk 2]
```diff
@@ -634,6 +634,33 @@ ptrdiff_t TClingClassInfo::GetBaseOffset(TClingClassInfo* base, void* address, b
...
+   cling::Interpreter::PushTransactionRAII RAII(fInterp);
+   const clang::DeclContext *DC = cast<DeclContext>(fDecl);
+   if (!DC)
+       return res;
+
+   clang::DeclContext::decl_iterator iter = DC->noload_decls_begin();
+   while (iter != DC->noload_decls_end()) {
```
Comment (vgvassilev) [inline core/metacling/src/TClingClassInfo.cxx]: Perhaps you want to take a look how clang does it [here](https://github.com/root-project/root/blob/master/interpreter/llvm/src/tools/clang/lib/Sema/SemaLookup.cpp#L1196-L1215).

**Comment by etejedor** [comment]
@Axel-Naumann can you have a look at the PR and the comments of @vgvassilev ? Thanks!

File: core/metacling/src/TClingClassInfo.cxx [hunk 3]
```diff
@@ -634,6 +634,33 @@ ptrdiff_t TClingClassInfo::GetBaseOffset(TClingClassInfo* base, void* address, b
...
 
+std::vector<std::string> TClingClassInfo::GetUsingNamespaces()
+{
+   // Find and return all 'using' declarations of namespaces.
+   std::vector<std::string> res;
+
+   R__LOCKGUARD(gInterpreterMutex);
+
+   cling::Interpreter::PushTransactionRAII RAII(fInterp);
+   const clang::DeclContext *DC = cast<DeclContext>(fDecl);
```
Comment (Axel-Naumann) [inline core/metacling/src/TClingClassInfo.cxx]: `const auto DC = dyn_cast<DeclContext>(fDecl));`? Else `fDecl` not being a `DeclContext` will crash.

File: core/metacling/src/TClingClassInfo.cxx (hunk 1 above)
Comment (Axel-Naumann) [reply core/metacling/src/TClingClassInfo.cxx]: We have a cast right after, and that should be a `dyn_cast` - see my comment below for the more common pattern.

File: core/metacling/src/TClingClassInfo.cxx [hunk 4]
```diff
@@ -634,6 +634,33 @@ ptrdiff_t TClingClassInfo::GetBaseOffset(TClingClassInfo* base, void* address, b
...
+   if (!DC)
+       return res;
+
+   clang::DeclContext::decl_iterator iter = DC->noload_decls_begin();
```
Comment (Axel-Naumann) [inline core/metacling/src/TClingClassInfo.cxx]: Why `noload`?

File: core/metacling/src/TClingClassInfo.cxx (hunk 2 above)
Comment (Axel-Naumann) [reply core/metacling/src/TClingClassInfo.cxx]: This is a limited feature... Not sure how much we want to try to replicate proper lookup...

**Comment by wlav** [comment]
@Axel-Naumann ,

"This isn't going to ever work - C++ lookup is waaay too complex. And even if it were to work, we would be duplicating a large amount of (complex) functionality, and would have to update it with future language standards."

Except that a) it does work today (I think you're truly missing how much cppyy has evolved in functionality from the limited bindings that PyROOT offers, with portable binaries on all major platforms to boot), and b) most of the new language features are resolved in a way that is invisible to the bindings. A case like this PR is very rare.

Neither this decl stuff nor the slowness of TClass and friends are material at this point (and simple to replace). What hurts cppyy much more, and what is taking up much more workaround code, is where ROOT/meta breaks things, like removing std::, or tries too hard to be smart, like with template instantiations or overload matching, where it doesn't even follow C++'s rules.

**Comment by Axel-Naumann** [comment]
> I think you're truly missing how much cppyy has evolved

Totally. That said you either re-implement lookup rules or they don't work properly, there's not a lot of middle ground... Templated constructors? ADL? Friended operators? What about spaceship? So much fun!

> where ROOT/meta breaks things, like removing std::, or tries too hard to be smart, like with template instantiations or overload matching, where it doesn't even follow C++'s rules.

How are we then not talking about the same issue? Is it because you believe that I blame cppyy, while I merely describe the behavior? I.e. yes exactly, and as I really don't feel like messing with ROOT's type names our best way out is lookup: pass the original name to clang, no name normalization, and see what it says. How's that not addressing these issues?

**Comment by wlav** [comment]
Templated constructors actually work since last week (except not yet on Mac due to a bug in TClingCallFunc; I needed them for shared_ptr), and yes they're a hack. Code is in repo, but not yet released. ADL will never work properly in full automatic mode, but is already covered most of the way. I'm having that discussion right now with a user who wants to turn Python tuples into std::tuples: the problem is underspecified on the Python side. So although it can be made to generally work, user-provided types are necessary to cover corner cases, e.g. by handing template arguments to std::make_tuple. Python's three-way comparison operator is only syntax, not semantics, so the spaceship would simply be broken down in Python's richcompare and the only real issue again is ROOT/meta giving wrong overloads. That said, I'm moving away from operator lookups: not only are the overloads often wrong, so I'm doing my own ADL, but also TClingCallFunc generates broken code on Mac and Windows. Rather, I want to use a generic comparison template and let SFINAE deal with it.

The difference between removing std::, giving wrong overloads, etc. and string <-> decl is that the former is a problem b/c ROOT/meta gives the _wrong_ results. Strings are sometimes clunky, but perfectly fine if they are correct.

Yes, letting clang parse names (as done for function template lookups already) is exactly what should happen. But the next step is to funnel the results to the Python mappings, not expose cppyy to decls or their intricate details. E.g. the converters and executor lookups should be a combination of strings and categories, just like it is now (except that the categories are derived from string parsing, like looking for ends_in('*')). Doing that selection based on decls as suggested here and in the e-mail thread with Gerhard makes no sense whatsoever: you'd have to maintain three times the same functionality.

I would want something like this:

`
opaque_handle h = what_is(parent, name)
type_category t = what_type(h)
qual_category q = what_qualifiers(h)
true_name = unqualified_true(h)
opaque_handle p = true_parent(h)
etc.
`

with opaque_handle possibly the decl, and with t and q some enums or flags, and that would get rid of the name parsing in converters/executors, but NOT the string-based name lookups there.

I do not need listings per se, not even for classes, but see GetAllCppNames(): something to feed to dir(), tab-completion, and help() remains needed for interactive use.

**Comment by pcanal** [comment]
Failure in filemerger is likely due to the time delay between the PR being merged and the corresponding update in roottest.

@phsft-bot build

**Comment by etejedor** [comment]
@phsft-bot build

**Pull request merged**